README.md
# websocket-reconnect

🚀 `websocket-reconnect-pro`。websocket 断开自动重连，基于原生`websocket`扩展实现、无依赖、TypeScript 编写、rollup 打包、单元测试。

[GitHub](https://github.com/chenchenwuai/websocket-reconnect)，如果帮助到了你，请给我一个 star~ 💖，如果你发现 bug，请尽情的提 issue 或者 pr，灰常感谢 💖

## 安装

[websocket-reconnect-pro](https://www.npmjs.com/package/websocket-reconnect-pro)

```bash
npm install websocket-reconnect-pro
```

## 使用

```javascript
// new WebsocketReconnect(url,protocols,options)

var url = "ws://192.168.1.100:8000";
var ws = new WebsocketReconnect(url, [], {
	minReconnectionDelay: 3000,
	reconnectionDelayGrowFactor: 1.1,
	enableHeartbeat: true,
	debug: true,
});
console.log(ws);

ws.onopen = function () {
	console.log("connect success");
	setTimeout(() => {
		ws.close();
	}, 3000);
};
ws.onclose = function (e) {
	console.log("onclose", e);
};
ws.onerror = function (e) {
	console.log("onerror", e);
};
ws.onmessage = function (e) {
	console.log(JSON.parse(e.data));
};
ws.onreconnect = function (e) {
	console.log("onreconnect", e);
};
```

## API

### 参数

#### `url`

url 是一个 websocket 地址字符串，或者是一个返回字符串的方法

```
url: string | (() => string) | (() => Promise<string>)
```

#### `protocols`

一个协议字符串或者一个包含协议字符串的数组。这些字符串用于指定子协议，这样单个服务器可以实现多个 WebSocket 子协议（例如，您可能希望一台服务器能够根据指定的协议（protocol）处理不同类型的交互）。如果不指定协议字符串，则假定为空字符串

#### `options`

| options 属性                | 说明                                                                       | 类型    | 可选值 | 默认值                       |
| --------------------------- | -------------------------------------------------------------------------- | ------- | ------ | ---------------------------- |
| WebSocket                   | 内部实例化的 WebSocket 类，默认为浏览器自带的`WebSocket`                   | Object  | -      | WebSocket                    |
| maxReconnectionDelay        | 最大重连时间间隔(毫秒)                                                     | number  | -      | 10000                        |
| minReconnectionDelay        | 最小重连时间间隔(毫秒)                                                     | number  | -      | 1000 + Math.random() \* 4000 |
| reconnectionDelayGrowFactor | 重连时间增长率,基数为 minReconnectionDelay,最大不超过 maxReconnectionDelay | number  | -      | 1.3                          |
| minUptime                   | -                                                                          | number  | -      | 5000                         |
| connectionTimeout           | 连接超时时间(毫秒)                                                         | number  | -      | 4000                         |
| maxRetries                  | 最大重连次数                                                               | number  | -      | Infinity                     |
| maxEnqueuedMessages         | 最大消息队列数量，重连后成功后会依次发送                                   | number  | -      | Infinity                     |
| autoSendQueueMessage        | 重连成功后是否自动发送队列中的消息                                         | boolean | -      | true                         |
| startClosed                 | 是否 new 之后不自动连接                                                    | boolean | -      | false                        |
| enableHeartbeat             | 是否开启心跳                                                               | boolean | -      | false                        |
| pingTimeout                 | 心跳发送时间间隔(毫秒)                                                     | number  | -      | 10000                        |
| pongTimeout                 | 心跳接受时间间隔(毫秒)                                                     | number  | -      | 10000                        |
| pingMsg                     | 心跳消息                                                                   | string  | -      | "\r\n"                       |
| outputPingMsg               | 心跳消息是否能被 onWCSStringMessage 方法输出                               | boolean | -      | false                        |
| debug                       | 开启 debug 模式                                                            | boolean | -      | false                        |

### 属性

除了 `WebSocket` 自带的[属性](https://developer.mozilla.org/en-US/docs/Web/API/WebSocket#properties)（`binaryType`,`readyState `等）以外，还有以下属性:

| 属性             | 说明                                                                                                                                       | 类型      | 备注       |
| ---------------- | ------------------------------------------------------------------------------------------------------------------------------------------ | --------- | ---------- |
| ws               | websocket 实例                                                                                                                             | WebSocket | 只读       |
| retryCount       | 当前失败重连的次数                                                                                                                         | Number    | 只读       |
| messageQueue     | 重连后需要发送的消息队列，当调用 send 方法时，内部的 websocket 还未建立，或者 websocket 的 readyState 不是 OPEN 时，会把消息 push 进此数组 | Array     | 可读、可写 |
| onreconnect      | 用于指定连接失败后的回调函数。                                                                                                             | Function  | 可读、可写 |
| onWebsocketError | 用于内部执行 new Websocket()时触发的错误回调函数。url 无法连接会触发此回调函数                                                             | Function  | 可读、可写 |

### 方法

除了 `WebSocket` 自带的[方法](https://developer.mozilla.org/en-US/docs/Web/API/WebSocket#methods)（`close`,`send `等）以外，还有以下属性:

| 方法      | 说明                                         | 参数 |
| --------- | -------------------------------------------- | ---- |
| reconnect | 主动调用重新连接方法，会重置 retryCount 为 0 | -    |

## 部分定义 .d.ts

```typescript
export default class WebsocketReconnect {
	private _ws?;
	private _binaryType;
	private _messageQueue;
	private _listeners;
	private _retryCount;
	private _uptimeTimeout;
	private _connectTimeout;
	private _shouldReconnect;
	private _connectLock;
	private _closeCalled;
	private _pingIntervalId;
	private _pongTimeoutId;
	private _nextConnectDelay;
	private readonly _url;
	private readonly _protocols;
	protected readonly _options: Required<Options>;
	constructor(
		url: UrlProvider,
		protocols?: string | string[],
		options?: Options
	);
	get binaryType(): BinaryType;
	set binaryType(value: BinaryType);
	get messageQueue(): Message[];
	set messageQueue(value: Message[]);
	get retryCount(): number;
	get bufferedAmount(): number;
	get extensions(): string;
	get protocol(): string;
	get readyState(): number;
	get url(): string;
	get ws(): WebSocket | undefined;
	static get CONNECTING(): number;
	static get OPEN(): number;
	static get CLOSING(): number;
	static get CLOSED(): number;
	get CONNECTING(): number;
	get OPEN(): number;
	get CLOSING(): number;
	get CLOSED(): number;
	onclose: ((event: CloseEvent) => any) | null;
	onerror: ((event: ErrorEvent) => any) | null;
	onmessage: ((message: MessageEvent) => void) | null;
	onopen: ((event: Event) => void) | null;
	onreconnect: ((options: ReconnectEventParams) => void) | null;
	onWebsocketError: ((error: unknown) => void) | null;
	close(code?: number, reason?: string): void;
	send(data: Message): void;
	addEventListener<T extends keyof WebSocketEventListenerMap>(
		type: T,
		listener: WebSocketEventListenerMap[T]
	): void;
	removeEventListener<T extends keyof WebSocketEventListenerMap>(
		type: T,
		listener: WebSocketEventListenerMap
	): void;
	dispatchEvent(event: Event): boolean;
	private _callEventListener;
	reconnect(code?: number, reason?: string): void;
	private _connect;
	private _disconnect;
	private _handleOpen;
	private _handleMessage;
	private _handleClose;
	private _handleError;
	private _handleReconnect;
	private _wait;
	private _getNextDelay;
	private _getNextUrl;
	private _handleTimeout;
	private _acceptOpen;
	private _debug;
	private _addEventListeners;
	private _removeListeners;
	private _heartCheck;
	private _heartStart;
	private _heartReset;
	private _clearTimeouts;
}
```

```typescript
export declare class Event {
	target: any;
	type: string;
	constructor(type: string, target: any);
}
export declare class ErrorEvent extends Event {
	message: string;
	error: Error;
	constructor(error: Error, target: any);
}
export declare class CloseEvent extends Event {
	code: number;
	reason: string;
	wasClean: boolean;
	constructor(
		code: number | undefined,
		reason: string | undefined,
		target: any
	);
}
export declare type ReconnectEventParams = {
	maxRetries: number;
	reconnectDelay: number;
	retryCount: number;
};
export interface WebSocketEventMap {
	close: CloseEvent;
	error: ErrorEvent;
	message: MessageEvent;
	open: Event;
	reconnect: ReconnectEventParams;
}
export interface WebSocketEventListenerMap {
	close: (event: CloseEvent) => void | {
		handleEvent: (event: CloseEvent) => void;
	};
	error: (event: ErrorEvent) => void | {
		handleEvent: (event: ErrorEvent) => void;
	};
	message: (event: MessageEvent) => void | {
		handleEvent: (event: MessageEvent) => void;
	};
	open: (event: Event) => void | {
		handleEvent: (event: Event) => void;
	};
	reconnect: (options: ReconnectEventParams) => void;
}
export declare type ListenersMap = {

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## 1.7.0 (2022-11-15)


### Features

* 增加 autoSendQueueMessage 选项 ([9dca31b](https://github.com/chenchenwuai/websocket-reconnect/commit/9dca31bac65dad5e8938999378dfb359a50095a5))
* 增加参数outputPingMsg ([8bd7b5b](https://github.com/chenchenwuai/websocket-reconnect/commit/8bd7b5be437c048d9d57d934e899c3f0a7185d28))
* 增加属性 ws ([15c4292](https://github.com/chenchenwuai/websocket-reconnect/commit/15c4292dcb02965df2c4d9cc47616ae701ba29f3))
* 增加心跳功能 ([cb4749c](https://github.com/chenchenwuai/websocket-reconnect/commit/cb4749c5b9749279a7073fa9962b032a679b1adb))
* 增加压缩功能 ([c554c42](https://github.com/chenchenwuai/websocket-reconnect/commit/c554c42344963e94555077504293bd3126785cd4))
* 增加重连回调函数；增加messageQueue属性 ([ce61b63](https://github.com/chenchenwuai/websocket-reconnect/commit/ce61b6380d8b689a94843c574739ef239cfb511a))
* 增加onWebsocketError 回调方法 ([2166ccf](https://github.com/chenchenwuai/websocket-reconnect/commit/2166ccf6eeb6ff1698fd86bec456a5736cbcf03a))


### Bug Fixes

* 修复测试未通过问题 ([0317cdf](https://github.com/chenchenwuai/websocket-reconnect/commit/0317cdfd1bce028c7136e3ee8ed6e73ecf2bacf9))
* 修复重连没有回调问题，和调整定义类型文件 ([e096d85](https://github.com/chenchenwuai/websocket-reconnect/commit/e096d85c9a4bc1da64ac3be1b30b04e40c8fecdb))
* **core:** 修复不能重连的bug ([48798a1](https://github.com/chenchenwuai/websocket-reconnect/commit/48798a1108d426357fa190151c00052106496bac))
* 修复options逻辑 ([86b5e6e](https://github.com/chenchenwuai/websocket-reconnect/commit/86b5e6eb6b3fd0f4007125ad980a48f9a11f73b3))
* **core:** 修复子类不能访问_options的问题 ([9847513](https://github.com/chenchenwuai/websocket-reconnect/commit/98475135f0161ed3f1ce71bfd8ddc366ae3d847a))

## [1.6.0](https://github.com/chenchenwuai/websocket-reconnect/compare/v1.5.0...v1.6.0) (2022-02-16)


### Features

* 增加onWebsocketError 回调方法 ([2166ccf](https://github.com/chenchenwuai/websocket-reconnect/commit/2166ccf6eeb6ff1698fd86bec456a5736cbcf03a))

## [1.5.0](https://github.com/chenchenwuai/websocket-reconnect/compare/v1.4.2...v1.5.0) (2022-02-16)


### Features

* 增加属性 ws ([15c4292](https://github.com/chenchenwuai/websocket-reconnect/commit/15c4292dcb02965df2c4d9cc47616ae701ba29f3))

### [1.4.2](https://github.com/chenchenwuai/websocket-reconnect/compare/v1.4.1...v1.4.2) (2021-12-10)

### [1.4.1](https://github.com/chenchenwuai/websocket-reconnect/compare/v1.4.0...v1.4.1) (2021-08-18)

## [1.4.0](https://github.com/chenchenwuai/websocket-reconnect/compare/v1.3.1...v1.4.0) (2021-08-18)


### Features

* 增加参数outputPingMsg ([8bd7b5b](https://github.com/chenchenwuai/websocket-reconnect/commit/8bd7b5be437c048d9d57d934e899c3f0a7185d28))

### [1.3.1](https://github.com/chenchenwuai/websocket-reconnect/compare/v1.3.0...v1.3.1) (2021-08-17)


### Bug Fixes

* 修复测试未通过问题 ([0317cdf](https://github.com/chenchenwuai/websocket-reconnect/commit/0317cdfd1bce028c7136e3ee8ed6e73ecf2bacf9))
* 修复重连没有回调问题，和调整定义类型文件 ([e096d85](https://github.com/chenchenwuai/websocket-reconnect/commit/e096d85c9a4bc1da64ac3be1b30b04e40c8fecdb))

## [1.3.0](https://github.com/chenchenwuai/websocket-reconnect/compare/v1.2.0...v1.3.0) (2021-08-17)


### Features

* 增加重连回调函数；增加messageQueue属性 ([ce61b63](https://github.com/chenchenwuai/websocket-reconnect/commit/ce61b6380d8b689a94843c574739ef239cfb511a))

## [1.2.0](https://github.com/chenchenwuai/websocket-reconnect/compare/v1.1.3...v1.2.0) (2021-03-11)


### Features

* 增加压缩功能 ([c554c42](https://github.com/chenchenwuai/websocket-reconnect/commit/c554c42344963e94555077504293bd3126785cd4))

### [1.1.3](https://github.com/chenchenwuai/websocket-reconnect/compare/v1.1.2...v1.1.3) (2021-03-11)


### Bug Fixes

* **core:** 修复不能重连的bug ([48798a1](https://github.com/chenchenwuai/websocket-reconnect/commit/48798a1108d426357fa190151c00052106496bac))

### [1.1.2](https://github.com/chenchenwuai/websocket-reconnect/compare/v1.1.1...v1.1.2) (2021-03-11)


### Bug Fixes

* 修复options逻辑 ([86b5e6e](https://github.com/chenchenwuai/websocket-reconnect/commit/86b5e6eb6b3fd0f4007125ad980a48f9a11f73b3))

### [1.1.1](https://github.com/chenchenwuai/websocket-reconnect/compare/v1.1.0...v1.1.1) (2021-03-10)


### Bug Fixes

* **core:** 修复子类不能访问_options的问题 ([9847513](https://github.com/chenchenwuai/websocket-reconnect/commit/98475135f0161ed3f1ce71bfd8ddc366ae3d847a))

## [1.1.0](https://github.com/chenchenwuai/websocket-reconnect/compare/v1.0.5...v1.1.0) (2021-03-10)


### Features


lib/websocket-reconnect.esm.js
/*!
 *  websocket-reconnect v1.3.1
 *  (c) 2020-2021 chenwuai
 * Released under the MIT License.
 */
/*! *****************************************************************************
Copyright (c) Microsoft Corporation.

Permission to use, copy, modify, and/or distribute this software for any
purpose with or without fee is hereby granted.

THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
PERFORMANCE OF THIS SOFTWARE.
***************************************************************************** */
/* global Reflect, Promise */

var extendStatics = function(d, b) {
    extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
    return extendStatics(d, b);
};

function __extends(d, b) {
    if (typeof b !== "function" && b !== null)
        throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
    extendStatics(d, b);
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
}

var __assign = function() {
    __assign = Object.assign || function __assign(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};

function __values(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
}

function __read(o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
}

/** @deprecated */
function __spread() {
    for (var ar = [], i = 0; i < arguments.length; i++)
        ar = ar.concat(__read(arguments[i]));
    return ar;
}

var Event = function () {
  function Event(type, target) {
    this.target = target;
    this.type = type;
  }

  return Event;
}();

var ErrorEvent = function (_super) {
  __extends(ErrorEvent, _super);

  function ErrorEvent(error, target) {
    var _this = _super.call(this, 'error', target) || this;

    _this.message = error.message;
    _this.error = error;
    return _this;
  }

  return ErrorEvent;
}(Event);

var CloseEvent = function (_super) {
  __extends(CloseEvent, _super);

  function CloseEvent(code, reason, target) {
    if (code === void 0) {
      code = 1000;
    }

    if (reason === void 0) {
      reason = '';
    }

    var _this = _super.call(this, 'close', target) || this;

    _this.wasClean = true;
    _this.code = code;
    _this.reason = reason;
    return _this;
  }

  return CloseEvent;
}(Event);

var getGlobalWebSocket = function getGlobalWebSocket() {
  if (typeof WebSocket !== 'undefined') {
    return WebSocket;
  }
};

var isWebSocket = function isWebSocket(w) {
  return typeof w !== 'undefined' && !!w && w.CLOSING === 2;
};

var WebsocketReconnect = function () {
  function WebsocketReconnect(url, protocols, options) {
    var _this = this;

    if (options === void 0) {
      options = {};
    }

    this._binaryType = 'blob';
    this._messageQueue = [];
    this._listeners = {
      error: [],
      message: [],
      open: [],
      close: [],
      reconnect: []
    };
    this._retryCount = -1;
    this._shouldReconnect = true;
    this._connectLock = false;
    this._closeCalled = false;
    this._pingIntervalId = 0;
    this._pongTimeoutId = 0;
    this._nextConnectDelay = 0;
    this.onclose = null;
    this.onerror = null;
    this.onmessage = null;
    this.onopen = null;
    this.onreconnect = null;

    this._handleOpen = function (event) {
      _this._debug('open event');

      var minUptime = _this._options.minUptime;
      clearTimeout(_this._connectTimeout);
      _this._uptimeTimeout = setTimeout(function () {
        return _this._acceptOpen();
      }, minUptime);

      if (_this._ws) {
        _this._ws.binaryType = _this._binaryType;
      }

      _this._options.enableHeartbeat && _this._heartCheck();

      _this._messageQueue.forEach(function (message) {
        var _a;

        return (_a = _this._ws) === null || _a === void 0 ? void 0 : _a.send(message);
      });

      _this._messageQueue = [];

      if (_this.onopen) {
        _this.onopen(event);
      }

      _this._listeners.open.forEach(function (listener) {
        return _this._callEventListener(event, listener);
      });
    };

    this._handleMessage = function (event) {
      _this._debug('message event');

      if (event.data === _this._options.pingMsg || event.data === '') {
        clearTimeout(_this._pongTimeoutId);

        _this._debug('pong');

        if (!_this._options.outputPingMsg) {
          return;
        }
      }

      if (_this.onmessage) {
        _this.onmessage(event);
      }

      _this._listeners.message.forEach(function (listener) {
        return _this._callEventListener(event, listener);
      });
    };

    this._handleClose = function (event) {
      _this._debug('close event');

      _this._clearTimeouts();

      if (_this._shouldReconnect) {
        _this._connect();
      }

      if (_this.onclose) {
        _this.onclose(event);
      }

      _this._listeners.close.forEach(function (listener) {
        return _this._callEventListener(event, listener);
      });
    };

    this._handleError = function (event) {
      _this._debug('error event', event.message);

      _this._disconnect(undefined, event.message === 'TIMEOUT' ? 'timeout' : undefined);

      if (_this.onerror) {
        _this.onerror(event);
      }

      _this._debug('exec error listeners');

      _this._listeners.error.forEach(function (listener) {
        return _this._callEventListener(event, listener);
      });

      _this._connect();
    };

    this._handleReconnect = function () {
      _this._debug('reconnect event');

      var maxRetries = _this._options.maxRetries;

      if (_this.onreconnect) {
        _this.onreconnect({
          maxRetries: maxRetries,
          reconnectDelay: _this._nextConnectDelay,
          retryCount: _this._retryCount
        });
      }

      _this._listeners.reconnect.forEach(function (listener) {
        return _this._callEventListener({
          maxRetries: maxRetries,
          reconnectDelay: _this._nextConnectDelay,
          retryCount: _this._retryCount
        }, listener);
      });
    };

    this._url = url;
    this._protocols = protocols !== null && protocols !== void 0 ? protocols : '';
    this._options = __assign({
      WebSocket: getGlobalWebSocket(),
      maxReconnectionDelay: 10000,
      minReconnectionDelay: 1000 + Math.random() * 4000,
      minUptime: 5000,
      reconnectionDelayGrowFactor: 1.3,
      connectionTimeout: 4000,
      maxRetries: Infinity,
      maxEnqueuedMessages: Infinity,
      startClosed: false,
      enableHeartbeat: false,
      pingTimeout: 10000,
      pongTimeout: 10000,
      pingMsg: '\r\n',
      outputPingMsg: false,
      debug: false
    }, options);

    if (this._options.startClosed) {
      this._shouldReconnect = false;
    }

    this._connect();
  }

  Object.defineProperty(WebsocketReconnect.prototype, "binaryType", {
    get: function get() {
      return this._ws ? this._ws.binaryType : this._binaryType;
    },
    set: function set(value) {
      this._binaryType = value;

      if (this._ws) {
        this._ws.binaryType = value;
      }
    },
    enumerable: false,
    configurable: true
  });
  Object.defineProperty(WebsocketReconnect.prototype, "messageQueue", {
    get: function get() {
      return this._messageQueue;
    },
    set: function set(value) {
      this._messageQueue = value;
    },
    enumerable: false,
    configurable: true
  });
  Object.defineProperty(WebsocketReconnect.prototype, "retryCount", {
    get: function get() {
      return Math.max(this._retryCount, 0);
    },
    enumerable: false,
    configurable: true
  });
  Object.defineProperty(WebsocketReconnect.prototype, "bufferedAmount", {
    get: function get() {
      var bytes = this._messageQueue.reduce(function (acc, message) {
        if (typeof message === 'string') {
          acc += message.length;
        } else if (message instanceof Blob) {
          acc += message.size;
        } else {
          acc += message.byteLength;
        }

        return acc;
      }, 0);

      return bytes + (this._ws ? this._ws.bufferedAmount : 0);
    },
    enumerable: false,
    configurable: true
  });
  Object.defineProperty(WebsocketReconnect.prototype, "extensions", {
    get: function get() {
      return this._ws ? this._ws.extensions : '';
    },
    enumerable: false,
    configurable: true
  });
  Object.defineProperty(WebsocketReconnect.prototype, "protocol", {
    get: function get() {
      return this._ws ? this._ws.protocol : '';
    },
    enumerable: false,
    configurable: true
  });
  Object.defineProperty(WebsocketReconnect.prototype, "readyState", {
    get: function get() {
      if (this._ws) {
        return this._ws.readyState;
      }

      return this._options.startClosed ? WebSocket.CLOSED : WebSocket.CLOSING;
    },
    enumerable: false,
    configurable: true
  });
  Object.defineProperty(WebsocketReconnect.prototype, "url", {
    get: function get() {
      return this._ws ? this._ws.url : '';
    },
    enumerable: false,
    configurable: true
  });
  Object.defineProperty(WebsocketReconnect, "CONNECTING", {
    get: function get() {
      return 0;
    },
    enumerable: false,
    configurable: true
  });
  Object.defineProperty(WebsocketReconnect, "OPEN", {
    get: function get() {
      return 1;
    },
    enumerable: false,
    configurable: true
  });
  Object.defineProperty(WebsocketReconnect, "CLOSING", {
    get: function get() {
      return 2;
    },
    enumerable: false,
    configurable: true
  });
  Object.defineProperty(WebsocketReconnect, "CLOSED", {
    get: function get() {
      return 3;
    },
    enumerable: false,
    configurable: true
  });
  Object.defineProperty(WebsocketReconnect.prototype, "CONNECTING", {
    get: function get() {
      return WebSocket.CONNECTING;
    },
    enumerable: false,
    configurable: true
  });
  Object.defineProperty(WebsocketReconnect.prototype, "OPEN", {
    get: function get() {
      return WebSocket.OPEN;
    },
    enumerable: false,
    configurable: true
  });
  Object.defineProperty(WebsocketReconnect.prototype, "CLOSING", {
    get: function get() {
      return WebSocket.CLOSING;
    },
    enumerable: false,
    configurable: true
  });
  Object.defineProperty(WebsocketReconnect.prototype, "CLOSED", {
    get: function get() {
      return WebSocket.CLOSED;
    },
    enumerable: false,
    configurable: true
  });

  WebsocketReconnect.prototype.close = function (code, reason) {
    this._closeCalled = true;
    this._shouldReconnect = false;

    this._clearTimeouts();

    if (!this._ws) {
      this._debug('close enqueued: no ws instance');

      return;
    }

    if (this._ws.readyState === this.CLOSED) {
      this._debug('close: already closed');

      return;
    }

    this._ws.close(code, reason);
  };

  WebsocketReconnect.prototype.send = function (data) {
    if (this._ws && this._ws.readyState === this.OPEN) {
      this._debug('send', data);

      this._ws.send(data);
    } else {
      var maxEnqueuedMessages = this._options.maxEnqueuedMessages;

      if (this._messageQueue.length < maxEnqueuedMessages) {
        this._debug('enqueue', data);

        this._messageQueue.push(data);
      }
    }
  };

  WebsocketReconnect.prototype.addEventListener = function (type, listener) {
    if (this._listeners[type]) {
      this._listeners[type].push(listener);
    }
  };

  WebsocketReconnect.prototype.removeEventListener = function (type, listener) {
    if (this._listeners[type]) {
      this._listeners = this._listeners[type].filter(function (l) {
        return l !== listener;
      });
    }
  };

  WebsocketReconnect.prototype.dispatchEvent = function (event) {
    var e_1, _a;

    var listeners = this._listeners[event.type];

    if (listeners) {
      try {
        for (var listeners_1 = __values(listeners), listeners_1_1 = listeners_1.next(); !listeners_1_1.done; listeners_1_1 = listeners_1.next()) {
          var listener = listeners_1_1.value;

          this._callEventListener(event, listener);
        }
      } catch (e_1_1) {
        e_1 = {
          error: e_1_1
        };
      } finally {
        try {
          if (listeners_1_1 && !listeners_1_1.done && (_a = listeners_1["return"])) _a.call(listeners_1);
        } finally {
          if (e_1) throw e_1.error;
        }
      }
    }

    return true;
  };

  WebsocketReconnect.prototype._callEventListener = function (event, listener) {
    if ('handleEvent' in listener) {
      listener.handleEvent(event);
    } else {
      listener(event);
    }
  };

  WebsocketReconnect.prototype.reconnect = function (code, reason) {
    this._shouldReconnect = true;
    this._closeCalled = false;
    this._retryCount = -1;

    if (!this._ws || this._ws.readyState === WebSocket.CLOSED) {
      this._connect();
    } else {
      this._disconnect(code, reason);

      this._connect();
    }
  };

  WebsocketReconnect.prototype._connect = function () {
    var _this = this;

    this._debug('connect', '_connectLock', this._connectLock, '_shouldReconnect', this._shouldReconnect);

    if (this._connectLock || !this._shouldReconnect) {
      return;
    }

    this._connectLock = true;
    var _a = this._options,
        maxRetries = _a.maxRetries,
        connectionTimeout = _a.connectionTimeout,
        WebSocket = _a.WebSocket;

    if (this._retryCount >= maxRetries) {
      this._debug('max retries reached', this._retryCount, '>=', maxRetries);

      return;
    }

    this._retryCount++;

    this._debug('connect times', this._retryCount);

    this._removeListeners();

    if (!isWebSocket(WebSocket)) {
      throw Error('No valid WebSocket class provided');
    }

    var delay = this._getNextDelay();

    this._nextConnectDelay = delay;

    this._debug('connect wait', delay);

    if (this._retryCount >= 1) {
      this._handleReconnect();
    }

    this._wait(delay).then(function () {
      return _this._getNextUrl(_this._url);
    }).then(function (url) {
      if (_this._closeCalled) {
        return;
      }

      _this._debug('connect', {
        url: url,
        protocols: _this._protocols
      });

      _this._ws = _this._protocols ? new WebSocket(url, _this._protocols) : new WebSocket(url);
      _this._ws.binaryType = _this._binaryType;
      _this._connectLock = false;

      _this._addEventListeners();

      _this._connectTimeout = setTimeout(function () {
        return _this._handleTimeout();
      }, connectionTimeout);
    });
  };

  WebsocketReconnect.prototype._disconnect = function (code, reason) {
    if (code === void 0) {
      code = 1000;
    }

    this._clearTimeouts();

    if (!this._ws) {
      return;
    }

    this._removeListeners();

    try {
      this._ws.close(code, reason);

      this._handleClose(new CloseEvent(code, reason, this));
    } catch (error) {
      throw Error(error);
    }
  };

  WebsocketReconnect.prototype._wait = function (delay) {
    if (delay === void 0) {
      delay = 0;
    }

    return new Promise(function (resolve) {
      setTimeout(function () {
        resolve();
      }, delay);
    });
  };

  WebsocketReconnect.prototype._getNextDelay = function () {
    var _a = this._options,
        reconnectionDelayGrowFactor = _a.reconnectionDelayGrowFactor,
        minReconnectionDelay = _a.minReconnectionDelay,
        maxReconnectionDelay = _a.maxReconnectionDelay;
    var delay = 0;

    if (this._retryCount > 0) {
      delay = minReconnectionDelay * Math.pow(reconnectionDelayGrowFactor, this._retryCount);

      if (delay > maxReconnectionDelay) {
        delay = maxReconnectionDelay;
      }
    }

    delay = Math.ceil(delay);

    this._debug('next delay', delay);

    return delay;
  };

  WebsocketReconnect.prototype._getNextUrl = function (urlProvider) {
    if (typeof urlProvider === 'string') {
      return Promise.resolve(urlProvider);
    } else if (typeof urlProvider === 'function') {
      var url = urlProvider();

      if (typeof url === 'string') {
        return Promise.resolve(url);
      }

      if (url.then !== undefined) {
        return url;
      }
    }

    throw Error('Invalid URL');
  };

  WebsocketReconnect.prototype._handleTimeout = function () {
    this._debug('timeout event');

    this._handleError(new ErrorEvent(Error('TIMEOUT'), this));
  };

  WebsocketReconnect.prototype._acceptOpen = function () {
    this._debug('accept open');

    this._retryCount = 0;
  };

  WebsocketReconnect.prototype._debug = function () {
    var args = [];

    for (var _i = 0; _i < arguments.length; _i++) {
      args[_i] = arguments[_i];
    }

    if (this._options.debug) {
      console.log.apply(console, __spread(['[WSR]'], args));
    }
  };

  WebsocketReconnect.prototype._addEventListeners = function () {
    if (!this._ws) {
      return;
    }

    this._debug('addListeners');

    this._ws.addEventListener('open', this._handleOpen);

    this._ws.addEventListener('message', this._handleMessage);

    this._ws.addEventListener('close', this._handleClose);

    this._ws.addEventListener('error', this._handleError);
  };

  WebsocketReconnect.prototype._removeListeners = function () {
    if (!this._ws) {
      return;
    }

    this._debug('removeListeners');

    this._ws.removeEventListener('open', this._handleOpen);

    this._ws.removeEventListener('message', this._handleMessage);

    this._ws.removeEventListener('close', this._handleClose);

    this._ws.removeEventListener('error', this._handleError);
  };

  WebsocketReconnect.prototype._heartCheck = function () {
    this._heartReset();

    this._heartStart();
  };

  WebsocketReconnect.prototype._heartStart = function () {
    var _this = this;

    var _a = this._options,
        enableHeartbeat = _a.enableHeartbeat,
        pingTimeout = _a.pingTimeout,
        pongTimeout = _a.pongTimeout,
        pingMsg = _a.pingMsg;

    if (!this._shouldReconnect && !enableHeartbeat) {
      return;
    }

    this._pingIntervalId = setInterval(function () {
      _this._debug('ping');

      _this.send(pingMsg);

      _this._pongTimeoutId = setTimeout(function () {
        _this._debug('Pong Timeout');

        _this.reconnect();
      }, pongTimeout);
    }, pingTimeout);
  };

  WebsocketReconnect.prototype._heartReset = function () {
    clearInterval(this._pingIntervalId);
    clearTimeout(this._pongTimeoutId);
  };

  WebsocketReconnect.prototype._clearTimeouts = function () {
    clearTimeout(this._connectTimeout);
    clearTimeout(this._uptimeTimeout);

    this._heartReset();
  };

  return WebsocketReconnect;
}();

export default WebsocketReconnect;
//# sourceMappingURL=websocket-reconnect.esm.js.map
